File: backend/routers/adaptations.py
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import HTMLResponse, Response
from sqlalchemy.orm import Session

from .. import models, schemas
from ..db import get_db
from ..deps import current_teacher
from ..services import adaptation, versioning

router = APIRouter(prefix="/api", tags=["adaptations"])


def _summary(v: models.LessonPlanVersion) -> schemas.VersionSummary:
    return schemas.VersionSummary(
        version_id=v.version_id,
        version_number=v.version_number,
        parent_version_id=v.parent_version_id,
        is_head=bool(v.is_head),
        instruction=v.instruction,
        model_used=v.model_used,
        provider=v.provider,
        token_count=v.token_count,
        created_at=v.created_at,
    )


def _adaptation_out(db: Session, adapted_id: int) -> schemas.AdaptationOut:
    head = versioning.head_version(db, adapted_id)
    if not head:
        raise HTTPException(404, "no versions for this adaptation")
    versions = versioning.list_versions(db, adapted_id)
    adapted = db.get(models.AdaptedLesson, adapted_id)
    return schemas.AdaptationOut(
        adapted_id=adapted.adapted_id,
        lesson_id=adapted.lesson_id,
        teacher_id=adapted.teacher_id,
        cluster_id=adapted.cluster_id,
        head_version=_summary(head),
        versions=[_summary(v) for v in versions],
    )


@router.post("/adapt", response_model=schemas.AdaptationOut)
def adapt(
    body: schemas.AdaptRequest,
    db: Session = Depends(get_db),
    teacher: models.Teacher = Depends(current_teacher),
) -> schemas.AdaptationOut:
    try:
        version = adaptation.generate(
            db,
            teacher=teacher,
            lesson_id=body.lesson_id,
            cluster_id=body.cluster_id,
            kb_ids=body.kb_ids,
            include_student_context=body.include_student_context,
        )
    except RuntimeError as e:
        raise HTTPException(400, str(e)) from e
    return _adaptation_out(db, version.adapted_id)


@router.post("/adaptations/{adapted_id}/refine", response_model=schemas.AdaptationOut)
def refine(
    adapted_id: int,
    body: schemas.RefineRequest,
    db: Session = Depends(get_db),
    teacher: models.Teacher = Depends(current_teacher),
) -> schemas.AdaptationOut:
    try:
        adaptation.refine(db, teacher=teacher, adapted_id=adapted_id, instruction=body.instruction)
    except LookupError as e:
        raise HTTPException(404, str(e)) from e
    except RuntimeError as e:
        raise HTTPException(400, str(e)) from e
    return _adaptation_out(db, adapted_id)


@router.get("/adaptations/{adapted_id}", response_model=schemas.AdaptationOut)
def get_adaptation(adapted_id: int, db: Session = Depends(get_db), teacher: models.Teacher = Depends(current_teacher)) -> schemas.AdaptationOut:
    adapted = db.get(models.AdaptedLesson, adapted_id)
    if not adapted:
        raise HTTPException(404, "not found")
    if adapted.teacher_id != teacher.teacher_id and teacher.role != "admin":
        raise HTTPException(403, "forbidden")
    return _adaptation_out(db, adapted_id)


@router.get("/adaptations/{adapted_id}/versions", response_model=list[schemas.VersionSummary])
def list_versions(adapted_id: int, db: Session = Depends(get_db), teacher: models.Teacher = Depends(current_teacher)) -> list[schemas.VersionSummary]:
    adapted = db.get(models.AdaptedLesson, adapted_id)
    if not adapted:
        raise HTTPException(404, "not found")
    if adapted.teacher_id != teacher.teacher_id and teacher.role != "admin":
        raise HTTPException(403, "forbidden")
    return [_summary(v) for v in versioning.list_versions(db, adapted_id)]


@router.get("/adaptations/{adapted_id}/versions/{version_id}", response_model=schemas.VersionDetail)
def get_version(adapted_id: int, version_id: int, db: Session = Depends(get_db), teacher: models.Teacher = Depends(current_teacher)) -> schemas.VersionDetail:
    adapted = db.get(models.AdaptedLesson, adapted_id)
    if not adapted:
        raise HTTPException(404, "not found")
    if adapted.teacher_id != teacher.teacher_id and teacher.role != "admin":
        raise HTTPException(403, "forbidden")
    v = db.get(models.LessonPlanVersion, version_id)
    if not v or v.adapted_id != adapted_id:
        raise HTTPException(404, "version not found")
    summary = _summary(v).model_dump()
    return schemas.VersionDetail(
        **summary,
        rendered_html=v.rendered_html,
        plan_json=versioning.parse_plan_json(v),
    )


@router.post("/adaptations/{adapted_id}/rollback", response_model=schemas.AdaptationOut)
def rollback(adapted_id: int, body: schemas.RollbackRequest, db: Session = Depends(get_db), teacher: models.Teacher = Depends(current_teacher)) -> schemas.AdaptationOut:
    adapted = db.get(models.AdaptedLesson, adapted_id)
    if not adapted:
        raise HTTPException(404, "not found")
    if adapted.teacher_id != teacher.teacher_id and teacher.role != "admin":
        raise HTTPException(403, "forbidden")
    try:
        versioning.rollback_to(db, adapted_id=adapted_id, version_id=body.version_id)
    except LookupError as e:
        raise HTTPException(404, str(e)) from e
    db.commit()
    return _adaptation_out(db, adapted_id)


@router.get("/adaptations/{adapted_id}/versions/{version_id}/print", response_class=HTMLResponse)
def print_version(adapted_id: int, version_id: int, db: Session = Depends(get_db), teacher: models.Teacher = Depends(current_teacher)) -> HTMLResponse:
    adapted = db.get(models.AdaptedLesson, adapted_id)
    if not adapted:
        raise HTTPException(404, "not found")
    if adapted.teacher_id != teacher.teacher_id and teacher.role != "admin":
        raise HTTPException(403, "forbidden")
    v = db.get(models.LessonPlanVersion, version_id)
    if not v or v.adapted_id != adapted_id:
        raise HTTPException(404, "version not found")
    return HTMLResponse(content=v.rendered_html)


@router.get("/adaptations/{adapted_id}/versions/{version_id}/export.html")
def export_version(adapted_id: int, version_id: int, db: Session = Depends(get_db), teacher: models.Teacher = Depends(current_teacher)) -> Response:
    adapted = db.get(models.AdaptedLesson, adapted_id)
    if not adapted:
        raise HTTPException(404, "not found")
    if adapted.teacher_id != teacher.teacher_id and teacher.role != "admin":
        raise HTTPException(403, "forbidden")
    v = db.get(models.LessonPlanVersion, version_id)
    if not v or v.adapted_id != adapted_id:
        raise HTTPException(404, "version not found")
    filename = f"adapt-lesson-{adapted_id}-v{v.version_number}.html"
    return Response(
        content=v.rendered_html,
        media_type="text/html",
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )


@router.post("/adaptations/{adapted_id}/feedback")
def submit_feedback(
    adapted_id: int,
    body: schemas.FeedbackIn,
    db: Session = Depends(get_db),
    teacher: models.Teacher = Depends(current_teacher),
) -> dict:
    adapted = db.get(models.AdaptedLesson, adapted_id)
    if not adapted:
        raise HTTPException(404, "not found")
    if adapted.teacher_id != teacher.teacher_id:
        raise HTTPException(403, "forbidden")
    fb = models.AdaptationFeedback(adapted_id=adapted_id, rating=body.rating, comments=body.comments)
    db.add(fb)
    db.commit()
    return {"ok": True, "feedback_id": fb.feedback_id}

File: server/src/routes/adaptations.js
const express = require('express');
const router = express.Router();
const requireAuth = require('../middleware/auth');
const adaptationService = require('../services/adaptation');
const versioningService = require('../services/versioning');
const db = require('../db');

function requireAdaptationOwner(req, res, next) {
  const adaptedId = parseInt(req.params.adapted_id);
  const adapted = db.prepare("SELECT teacher_id FROM adapted_lesson WHERE adapted_id = ?").get(adaptedId);
  if (!adapted) return res.status(404).json({ error: 'Adaptation not found' });
  if (adapted.teacher_id !== req.user.teacher_id && req.user.role !== 'admin') {
    return res.status(403).json({ error: 'Access denied' });
  }
  req.adapted = adapted;
  next();
}

function versionSummary(v) {
  return {
    version_id: v.version_id,
    version_number: v.version_number,
    parent_version_id: v.parent_version_id,
    is_head: Boolean(v.is_head),
    instruction: v.instruction,
    model_used: v.model_used,
    provider: v.provider,
    token_count: v.token_count,
    created_at: v.created_at,
  };
}

function adaptationOut(adaptedId) {
  const head = versioningService.headVersion(adaptedId);
  if (!head) return { error: 'no versions for this adaptation' };
  const versions = versioningService.listVersions(adaptedId);
  const adapted = db.prepare("SELECT * FROM adapted_lesson WHERE adapted_id = ?").get(adaptedId);
  return {
    adapted_id: adapted.adapted_id,
    lesson_id: adapted.lesson_id,
    teacher_id: adapted.teacher_id,
    cluster_id: adapted.cluster_id,
    head_version: versionSummary(head),
    versions: versions.map(versionSummary),
  };
}

// POST /api/adapt — Generate adapted lesson plan
router.post('/adapt', requireAuth, async (req, res) => {
  const { lesson_id, cluster_id, kb_ids, include_student_context } = req.body;
  if (!lesson_id || !cluster_id) return res.status(400).json({ error: 'lesson_id and cluster_id required' });
  try {
    const result = await adaptationService.generate({
      teacherId: req.user.teacher_id,
      lessonId: lesson_id,
      clusterId: cluster_id,
      kbIds: kb_ids || [],
      includeStudentContext: include_student_context !== false,
    });
    const out = adaptationOut(result.adaptedId);
    if (out.error) return res.status(404).json({ error: out.error });
    res.status(201).json(out);
  } catch (e) {
    if (e.message.includes('No LLM configured')) return res.status(400).json({ error: e.message });
    res.status(500).json({ error: e.message });
  }
});

// POST /api/adaptations/:adapted_id/refine — Refine with feedback
router.post('/adaptations/:adapted_id/refine', requireAuth, requireAdaptationOwner, async (req, res) => {
  const { instruction } = req.body;
  if (!instruction) return res.status(400).json({ error: 'instruction required' });
  try {
    await adaptationService.refine({
      teacherId: req.user.teacher_id,
      adaptedId: parseInt(req.params.adapted_id),
      instruction,
    });
    res.json(adaptationOut(parseInt(req.params.adapted_id)));
  } catch (e) {
    if (e.message.includes('not found')) return res.status(404).json({ error: e.message });
    if (e.message.includes('No LLM configured')) return res.status(400).json({ error: e.message });
    res.status(500).json({ error: e.message });
  }
});

// GET /api/adaptations/:adapted_id — Get adaptation with head version
router.get('/adaptations/:adapted_id', requireAuth, requireAdaptationOwner, (req, res) => {
  const out = adaptationOut(parseInt(req.params.adapted_id));
  if (out.error) return res.status(404).json({ error: out.error });
  res.json(out);
});

// GET /api/adaptations/:adapted_id/versions — List all versions
router.get('/adaptations/:adapted_id/versions', requireAuth, requireAdaptationOwner, (req, res) => {
  const versions = versioningService.listVersions(parseInt(req.params.adapted_id));
  res.json(versions.map(versionSummary));
});

// GET /api/adaptations/:adapted_id/versions/:version_id — Get version detail
router.get('/adaptations/:adapted_id/versions/:version_id', requireAuth, requireAdaptationOwner, (req, res) => {
  const v = db.prepare("SELECT * FROM lesson_plan_version WHERE version_id = ? AND adapted_id = ?").get(
    parseInt(req.params.version_id), parseInt(req.params.adapted_id)
  );
  if (!v) return res.status(404).json({ error: 'version not found' });
  res.json({
    ...versionSummary(v),
    rendered_html: v.rendered_html,
    plan_json: versioningService.parsePlanJson(v),
  });
});

// POST /api/adaptations/:adapted_id/rollback — Rollback to previous version
router.post('/adaptations/:adapted_id/rollback', requireAuth, requireAdaptationOwner, (req, res) => {
  const { version_id } = req.body;
  if (!version_id) return res.status(400).json({ error: 'version_id required' });
  try {
    versioningService.rollbackTo(parseInt(req.params.adapted_id), version_id);
    res.json(adaptationOut(parseInt(req.params.adapted_id)));
  } catch (e) {
    if (e.message.includes('not found')) return res.status(404).json({ error: e.message });
    res.status(500).json({ error: e.message });
  }
});

// GET /api/adaptations/:adapted_id/versions/:version_id/print — Print as HTML
router.get('/adaptations/:adapted_id/versions/:version_id/print', requireAuth, requireAdaptationOwner, (req, res) => {
  const v = db.prepare("SELECT rendered_html FROM lesson_plan_version WHERE version_id = ? AND adapted_id = ?").get(
    parseInt(req.params.version_id), parseInt(req.params.adapted_id)
  );
  if (!v) return res.status(404).json({ error: 'version not found' });
  res.setHeader('Content-Type', 'text/html; charset=utf-8');
  res.send(v.rendered_html);
});

// GET /api/adaptations/:adapted_id/versions/:version_id/export.html — Download as HTML file
router.get('/adaptations/:adapted_id/versions/:version_id/export.html', requireAuth, requireAdaptationOwner, (req, res) => {
  const v = db.prepare("SELECT rendered_html, version_number FROM lesson_plan_version WHERE version_id = ? AND adapted_id = ?").get(
    parseInt(req.params.version_id), parseInt(req.params.adapted_id)
  );
  if (!v) return res.status(404).json({ error: 'version not found' });
  const filename = `adapt-lesson-${req.params.adapted_id}-v${v.version_number}.html`;
  res.setHeader('Content-Disposition', `attachment; filename="${filename}"`);
  res.setHeader('Content-Type', 'text/html; charset=utf-8');
  res.send(v.rendered_html);
});

// POST /api/adaptations/:adapted_id/feedback — Submit feedback
router.post('/adaptations/:adapted_id/feedback', requireAuth, (req, res) => {
  const adaptedId = parseInt(req.params.adapted_id);
  const adapted = db.prepare("SELECT teacher_id FROM adapted_lesson WHERE adapted_id = ?").get(adaptedId);
  if (!adapted) return res.status(404).json({ error: 'Adaptation not found' });
  if (adapted.teacher_id !== req.user.teacher_id) return res.status(403).json({ error: 'Access denied' });

  const { rating, comments } = req.body;
  if (!rating || rating < 1 || rating > 5) return res.status(400).json({ error: 'rating must be 1-5' });

  const result = db.prepare(
    "INSERT INTO adaptation_feedback (adapted_id, rating, comments) VALUES (?, ?, ?)"
  ).run(adaptedId, rating, comments || null);

  res.json({ ok: true, feedback_id: result.lastInsertRowid });
});

module.exports = router;
